pennylane/qnn/torch.py

```diff
@@ -0,0 +1,311 @@
+    """
+
+    def __init__(
+        self, qnode, weight_shapes: dict, output_dim, init_method: Optional[Callable] = None
+    ):
+        if not TORCH_IMPORTED:
+            raise ImportError("TorchLayer requires PyTorch")
+        super().__init__()
+
+        self.sig = qnode.func.sig
+
+        if self.input_arg not in self.sig:
+            raise TypeError(
+                "QNode must include an argument with name {} for inputting data".format(
+                    self.input_arg
+                )
+            )
+
+        if self.input_arg in set(weight_shapes.keys()):
+            raise ValueError(
+                "{} argument should not have its dimension specified in "
+                "weight_shapes".format(self.input_arg)
+            )
+
+        if set(weight_shapes.keys()) | {self.input_arg} != set(self.sig.keys()):
+            raise ValueError("Must specify a shape for every non-input parameter in the QNode")
+
+        if qnode.func.var_pos:
+            raise TypeError("Cannot have a variable number of positional arguments")
+
+        if qnode.func.var_keyword:
+            raise TypeError("Cannot have a variable number of keyword arguments")
+
+        self.qnode = to_torch(qnode)
+        weight_shapes = {
+            weight: (tuple(size) if isinstance(size, Iterable) else (size,) if size > 1 else ())
+            for weight, size in weight_shapes.items()
+        }
+
+        # Allows output_dim to be specified as an int, e.g., 5, or as a length-1 tuple, e.g., (5,)
+        self.output_dim = output_dim[0] if isinstance(output_dim, Iterable) else output_dim
+
+        defaults = {
+            name for name, sig in self.sig.items() if sig.par.default != inspect.Parameter.empty
+        }
+        self.input_is_default = self.input_arg in defaults
+        if defaults - {self.input_arg} != set():
+            raise TypeError(
+                "Only the argument {} is permitted to have a default".format(self.input_arg)
+            )
+
+        if not init_method:
+            init_method = functools.partial(torch.nn.init.uniform_, b=2 * math.pi)
+
+        self.qnode_weights = {}
+        for name, size in weight_shapes.items():
+            if len(size) == 0:
+                self.qnode_weights[name] = torch.nn.Parameter(init_method(torch.Tensor(1))[0])
+            else:
+                self.qnode_weights[name] = torch.nn.Parameter(init_method(torch.Tensor(*size)))
+            self.register_parameter(name, self.qnode_weights[name])
+
+    def forward(self, inputs):  # pylint: disable=arguments-differ
+        """Evaluates a forward pass through the QNode based upon input data and the initialized
+        weights.
+
+        Args:
+            inputs (tensor): data to be processed
+
+        Returns:
+            tensor: output data
+        """
+        if len(inputs.shape) == 1:
+            return self._evaluate_qnode(inputs)
+
+        return torch.stack([self._evaluate_qnode(x) for x in inputs])
```

Comment:
TypeError: can't convert np.ndarray of type numpy.object_. The only supported types are: float64, float32, float16, int64, int32, int16, int8, uint8, and bool.

^ Good to know, this is another 'gotcha' we should keep in mind in case of tech support inquiries. Namely, the Torch interface doesn't work for QNodes with 'ragged' output.